orttraining/orttraining/core/optimizer/memory_alleviation.cc

```diff
@@ -0,0 +1,460 @@
+    if (candidate_output_args_map.find(&node) == candidate_output_args_map.end()) {
+      continue;
+    }
+
+    InlinedVector<const Node*> nodes_in_topological_order;
+    std::ostringstream oss;
+    std::ostringstream node_type_in_topological_order;
+    memory_alleviation::AlleviationType alleviation_type;
+    ORT_RETURN_IF_ERROR(SelectSubgraph(graph, node, fw_op_output_arg_used_map, nodes_in_topological_order,
+                                       oss, node_type_in_topological_order, alleviation_type));
+    if (nodes_in_topological_order.size() == 0) {
+      continue;
+    }
+
+    std::string postfix_if_any;
+    if (oss.tellp() != std::streampos(0)) {
+      postfix_if_any = " with its precedent nodes: " + oss.str();
+    }
+
+    LOGS(logger, WARNING) << "Node " << node.Name() << "(" << node.OpType() << ") can be recomputed" << postfix_if_any
+                          << "\n";
+
+    Node* replacement_node = nullptr;
+
+    bool modify_graph = (alleviation_type != memory_alleviation::AlleviationType::None);
+
+    if (modify_graph) {
+      ORT_RETURN_IF_ERROR(CreateRecomputeGraph(graph, nodes_in_topological_order, replacement_node));
+      ORT_ENFORCE(replacement_node);
+    }
+
+    std::string node_type_in_topological_order_str = node_type_in_topological_order.str();
+    subgraph_str_to_alleviation_type[node_type_in_topological_order_str] = alleviation_type;
+
+    for (size_t output_index : candidate_output_args_map[&node]) {
+      if (modify_graph) {
+        // Collect output edges (connecting to backward ops), to remove.
+        std::vector<graph_utils::GraphEdge> output_edges;
+        for (auto it = node.OutputEdgesBegin(), end = node.OutputEdgesEnd(); it != end; ++it) {
+          size_t src_output_idx = static_cast<size_t>(it->GetSrcArgIndex());
+          if (src_output_idx != output_index) {
+            continue;
+          }
+
+          if (!is_forward_op_map[it->GetNode().Index()]) {
+            // Remove the edge only connecting to backward op.
+            output_edges.push_back(graph_utils::GraphEdge::CreateGraphEdge(node, *it, false));
+          }
+        }
+
+        if (!output_edges.empty()) {
+          // Remove the output edges of the node first
+          graph_utils::GraphEdge::RemoveGraphEdges(graph, output_edges);
+
+          // Create connections between the replacement node and the outgoing nodes.
+          for (const auto& output_edge : output_edges) {
+            // Add new edge connecting the input with the output nodes directly.
+            // This also updates the destination node's input node args
+            graph.AddEdge(replacement_node->Index(), output_edge.dst_node, static_cast<int>(output_index),
+                          output_edge.dst_arg_index);
+          }
+        }
+      }
+
+      auto shape_str = TensorShapeProtoToString(node.OutputDefs()[output_index]->Shape());
+      stashed_activation_statistics[node_type_in_topological_order_str][shape_str]++;
```

Comment:
Shall we create the statistics and print them only when a certain log level is set? It will affect initialization time and not run time because this is part of the transformer.

Reply:
I have changeD the log level to INFO or verbose.